### PR TITLE
ci: Add pipeline change detection

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

part of https://github.com/envoyproxy/envoy/issues/23758

### CODE REVIEW

File: .azure-pipelines/pipelines.yml [hunk 1]
```diff
@@ -25,6 +25,73 @@ variables:
   value: $[or(eq(variables['isMain'], 'true'), eq(variables['isReleaseBranch'], 'true'))]
 
 stages:
+- stage: change_detection
```
Comment (phlax) [inline .azure-pipelines/pipelines.yml:28]: perhaps rename this to `environment` as we can also use it to detect eg VERSION

File: .azure-pipelines/pipelines.yml [hunk 2]
```diff
@@ -25,6 +25,73 @@ variables:
...
+  # dependsOn: ["change_detection"]
+  #
+  # To use a variable as a condition for a `stage` you can specify it as follows:
+  #
+  # condition: ne(dependencies.change_detection.outputs['repo.changed.mobileOnly'], 'true')
+  #
+  # To use a variable as a condition for a `job` you can specify it as follows:
+  #
+  # job:
+  #   condition: eq(stageDependencies.change_detection.repo.outputs['changed.mobileOnly'], 'true')
```
Comment (phlax) [inline .azure-pipelines/pipelines.yml:42]: ne

**Comment by alyssawilk** [comment]
hah I was about to ask if it was easier for you to pick this up than suggest all the fixes.  Thanks so much for tackling it!  LMK when this is ready as I'd love to get something in this week

**Comment by phlax** [comment]
i have not added any actual exclusions to this PR - just the framework for detecting changes - as when we add any exclusions we need to think carefully about how it will run on `main` or other release branches and on tagged commits/releases

im inclined to not exclude anything when a PR lands - this does risk PRs passing and then failing when they land - but i would rather that than commits passing and then releases not passing and i think we need to ensure every change is tested _somewhere_

**Comment by phlax** [comment]
testing PR is #24204

**Comment by phlax** [comment]
this is probably gtg altho we may need to iterate - testing is hard because the PR introduces change itself

**Comment by alyssawilk** [comment]
I'd be inclined to land with at least one CI we think is affected (i.e. https://github.com/envoyproxy/envoy/pull/24204/files) as I'm not sure landing this as-is is going to tell us anything. WDYT?

**Comment by phlax** [comment]
> WDYT?

adding a change here is probably not going to help as the PR contains change itself

i think land this and then run the test PR a couple of times, and then clarify what should/not be excluded

**Comment by phlax** [comment]
> adding a change here is probably not going to help as the PR contains change itself

hmm, actually thinking about it - yep i see - we need at least one thing to be excluded to test - ill add ...

**Comment by alyssawilk** [comment]
yeah I'll also say now we have this (or almost) I am so excited to say not run all the Envoy tests for docs only changes.  This is going to be pretty awesome :-D

**Comment by phlax** [comment]
k - ive added a condition which hopefully will only run coverage iff either:

- commit is postsubmit (on any branch)
- the PR is neither docs-only nor mobile-only

should work

**Review by alyssawilk — APPROVED** [review]
Let's see how it goes :-)

**Pull request merged**